src/core/border_router/routing_manager.cpp

```diff
@@ -3451,6 +3384,66 @@ Nat64::State RoutingManager::Nat64PrefixManager::GetState(void) const
 
 #endif // OPENTHREAD_CONFIG_NAT64_BORDER_ROUTING_ENABLE
 
+//---------------------------------------------------------------------------------------------------------------------
+// RaInfo
+
+void RoutingManager::RaInfo::IncrementTxCountAndSaveHash(const InfraIf::Icmp6Packet &aRaMessage)
+{
+    mTxCount++;
+    mLastHashIndex++;
+
+    if (mLastHashIndex == kNumHashEntries)
+    {
+        mLastHashIndex = 0;
+    }
+
+    CalculateHash(aRaMessage, mHashes[mLastHashIndex]);
+}
+
+bool RoutingManager::RaInfo::IsRaFromManager(const Ip6::Nd::RouterAdvert::RxMessage &aRaMessage) const
+{
+    // Determines whether or not a received RA message was prepared by
+    // by `RoutingManager` itself (is present in the saved `mHashes`).
+
+    bool     isFromManager = false;
+    uint16_t hashIndex     = mLastHashIndex;
+    uint32_t count         = Min<uint32_t>(mTxCount, kNumHashEntries);
+    Hash     hash;
+
+    CalculateHash(aRaMessage.GetAsPacket(), hash);
```

Comment:
I know we didn't before, but should we include IPv6 header information (e.g. IPv6 addresses) in the hash calculation as well?

Reply:
Not sure about this. We don't actually prepare the IPv6 header in `RoutingManager` (we don't know the host's  infra-if link-local address). We expect the IPv6 header to be prepared by the platform layer.  The `otPlatInfraIfSendIcmp6Nd()` expects an ICMPv6 message as its input and will then prepare the IPv6 header.